File: src/config/presets.ts
/**
 * 供应商目录（presets）加载层。
 *
 * 加载优先级（评审⑤）：
 *   1) 用户目录 ~/.cc-mini/presets.json（可选，用户加供应商不必动 node_modules）
 *   2) 包内随发布的 presets.json（dist 的上级目录）
 *   3) 内置常量 BUILTIN_PRESETS（最后兜底，等价于现版 $BuiltinPresetsJson）
 *
 * 任一步解析失败/格式不对就跌落到下一步，绝不抛错中断启动。
 */
import { existsSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';

import { resolveStorePaths } from './store.js';
import type { Preset, PresetEnv, PresetModels, PresetUrl } from './types.js';

export type { Preset } from './types.js';

/** 内置兜底目录（镜像仓库 presets.json）。第三方供应商**绝不**预置 `[1m]`（见 plan §3.1.1）。 */
export const BUILTIN_PRESETS: Preset[] = [
  {
    name: 'DeepSeek',
    auth: 'AUTH_TOKEN',
    effort: 'max',
    urls: [{ label: 'Anthropic 兼容', url: 'https://api.deepseek.com/anthropic' }],
    models: { opus: 'deepseek-v4-pro', sonnet: 'deepseek-v4-pro', haiku: 'deepseek-v4-flash' },
    modelsApi: 'https://api.deepseek.com/models', // Anthropic 兼容端点未实现 /v1/models，模型列表在 OpenAI 风格端点
    models1m: ['deepseek-v4-pro'],
  },
  {
    name: '智谱GLM',
    auth: 'AUTH_TOKEN',
    urls: [{ label: 'Anthropic 兼容', url: 'https://open.bigmodel.cn/api/anthropic' }],
    models: { opus: 'GLM-4.7', sonnet: 'GLM-4.7', haiku: 'glm-4.5-air' },
    env: {
      CLAUDE_CODE_DISABLE_NONESSENTIAL_TRAFFIC: '1',
    },
    modelsApi: 'https://open.bigmodel.cn/api/anthropic/v1/models',
    models1m: ['glm-5.2'],
  },
  {
    name: '小米MiMo',
    auth: 'AUTH_TOKEN',
    urls: [
      { label: '按量付费API', url: 'https://api.xiaomimimo.com/anthropic' },
      { label: 'TokenPlan', url: 'https://token-plan-cn.xiaomimimo.com/anthropic' },
    ],
    models: { opus: 'mimo-v2.5-pro', sonnet: 'mimo-v2.5-pro', haiku: 'mimo-v2.5-pro' },
    // MiMo 的 models 端点是 OpenAI 风格 /v1/models（不带 /anthropic 前缀），按 base 前缀匹配各自端点
    modelsApiMap: {
      'https://api.xiaomimimo.com/anthropic': 'https://api.xiaomimimo.com/v1/models',
      'https://token-plan-cn.xiaomimimo.com/anthropic': 'https://token-plan-cn.xiaomimimo.com/v1/models',
    },
    models1m: ['mimo-v2.5-pro'],
  },
  {
    name: 'Kimi (月之暗面)',
    auth: 'AUTH_TOKEN',
    effort: 'max',
    urls: [{ label: 'Anthropic 兼容', url: 'https://api.moonshot.cn/anthropic' }],
    models: { opus: 'kimi-k3', sonnet: 'kimi-k3', haiku: 'kimi-k3', fable: 'kimi-k3' },
    models1m: ['kimi-k3'],
  },
  {
    name: 'MiniMax',
    auth: 'AUTH_TOKEN',
    urls: [
      { label: '国内站', url: 'https://api.minimaxi.com/anthropic' },
      { label: '国际站', url: 'https://api.minimax.io/anthropic' },
    ],
    models: { opus: 'MiniMax-M3', sonnet: 'MiniMax-M3', haiku: 'MiniMax-M3', fable: 'MiniMax-M3' },
    models1m: ['MiniMax-M3'],
  },
  {
    name: 'OpenRouter',
    auth: 'API_KEY',
    urls: [{ label: 'Anthropic 兼容', url: 'https://openrouter.ai/api' }],
    models: { opus: 'anthropic/claude-opus-5', sonnet: 'anthropic/claude-sonnet-5', haiku: 'anthropic/claude-haiku-4-5', fable: 'anthropic/claude-fable-5' },
  },
  {
    name: '百度千帆',
    auth: 'AUTH_TOKEN',
    urls: [{ label: 'Coding Plan', url: 'https://qianfan.baidubce.com/anthropic/coding' }],
    models: { opus: 'qianfan-code-latest', sonnet: 'qianfan-code-latest', haiku: 'qianfan-code-latest', fable: 'qianfan-code-latest' },
    env: {
      CLAUDE_CODE_DISABLE_NONESSENTIAL_TRAFFIC: '1',
    },
  },
  {
    name: '阿里百炼',
    auth: 'AUTH_TOKEN',
    urls: [
      { label: 'Token Plan（个人/团队）', url: 'https://token-plan.cn-beijing.maas.aliyuncs.com/apps/anthropic' },
      { label: '按量付费', url: 'https://dashscope.aliyuncs.com/apps/anthropic' },
    ],
    models: { opus: 'qwen3.8-max', sonnet: 'qwen3.8-max', haiku: 'qwen3.6-flash', fable: 'qwen3.8-max' },
    models1m: ['qwen3.8-max', 'qwen3.7-max', 'qwen3.7-plus'],
  },
  {
    name: '火山方舟',
    auth: 'AUTH_TOKEN',
    urls: [{ label: 'Coding Plan', url: 'https://ark.cn-beijing.volces.com/api/coding' }],
    models: { opus: 'doubao-seed-2.0-code', sonnet: 'doubao-seed-2.0-code', haiku: 'doubao-seed-2.0-code', fable: 'doubao-seed-2.0-code' },
  },
  {
    name: '官方Anthropic',
    auth: 'API_KEY',
    urls: [{ label: '(留空，用登录态)', url: '' }],
    models: {},
  },
];

const asString = (v: unknown): string => (typeof v === 'string' ? v : '');

function normalizeUrl(raw: unknown): PresetUrl {
  const u = (raw && typeof raw === 'object' ? raw : {}) as Record<string, unknown>;
  return { label: asString(u.label), url: asString(u.url) };
}

function normalizeModels(raw: unknown): PresetModels {
  const m = (raw && typeof raw === 'object' ? raw : {}) as Record<string, unknown>;
  const models: PresetModels = {};
  if (typeof m.opus === 'string') models.opus = m.opus;
  if (typeof m.sonnet === 'string') models.sonnet = m.sonnet;
  if (typeof m.haiku === 'string') models.haiku = m.haiku;
  if (typeof m.fable === 'string') models.fable = m.fable;
  return models;
}

const PRESET_ENV_KEYS = ['CLAUDE_CODE_DISABLE_NONESSENTIAL_TRAFFIC', 'CLAUDE_CODE_SUBAGENT_MODEL'] as const;

function normalizeEnv(raw: unknown): PresetEnv | undefined {
  const m = (raw && typeof raw === 'object' ? raw : {}) as Record<string, unknown>;
  const env: PresetEnv = {};
  for (const key of PRESET_ENV_KEYS) {
    if (typeof m[key] === 'string') env[key] = m[key];
  }
  return Object.keys(env).length > 0 ? env : undefined;
}

function normalizePreset(raw: unknown): Preset | undefined {
  const p = (raw && typeof raw === 'object' ? raw : {}) as Record<string, unknown>;
  const name = asString(p.name).trim();
  if (!name) return undefined; // 无名条目直接丢弃
  const auth: Preset['auth'] = p.auth === 'API_KEY' ? 'API_KEY' : 'AUTH_TOKEN';
  const urls = Array.isArray(p.urls) ? p.urls.map(normalizeUrl) : [];
  const preset: Preset = { name, auth, urls, models: normalizeModels(p.models) };
  if (typeof p.effort === 'string' && p.effort.trim()) preset.effort = p.effort.trim();
  const env = normalizeEnv(p.env);
  if (env) preset.env = env;
  if (typeof p.models_api === 'string' && p.models_api.trim()) preset.modelsApi = p.models_api.trim();
  if (p.models_api_map && typeof p.models_api_map === 'object') {
    const map: Record<string, string> = {};
    for (const [k, v] of Object.entries(p.models_api_map as Record<string, unknown>)) {
      if (typeof v === 'string' && v.trim()) map[k.trim()] = v.trim();
    }
    if (Object.keys(map).length > 0) preset.modelsApiMap = map;
  }
  if (Array.isArray(p.models_1m)) {
    const list = p.models_1m.map(asString).map((s) => s.trim()).filter(Boolean);
    if (list.length > 0) preset.models1m = list;
  }
  return preset;
}

/** 把任意解析结果规整为 Preset[]；非数组或全空则返回 undefined（让调用方跌落兜底）。 */
function normalizePresets(raw: unknown): Preset[] | undefined {
  if (!Array.isArray(raw)) return undefined;
  const list = raw.map(normalizePreset).filter((p): p is Preset => p !== undefined);
  return list.length > 0 ? list : undefined;
}

/** 尝试读并解析一个 presets.json；任何问题都安静返回 undefined。 */
function tryLoadFile(file: string): Preset[] | undefined {
  try {
    if (!existsSync(file)) return undefined;
    return normalizePresets(JSON.parse(readFileSync(file, 'utf-8')));
  } catch {
    return undefined;
  }
}

/** 定位包内随发布的 presets.json（dist 的上级 = 包根；dev 下 = 仓库根）。 */
function packagePresetsFile(): string {
  // 本模块编译后位于 dist/config/presets.js → ../../presets.json = 包根。
  return fileURLToPath(new URL('../../presets.json', import.meta.url));
}

/**
 * 加载供应商目录。`storeDir` 来自 --store-dir（测试用），决定用户覆盖文件的位置。
 * 优先级见文件头注释。
 */
export function loadPresets(storeDir?: string): Preset[] {
  const userFile = join(resolveStorePaths(storeDir).dir, 'presets.json');
  return tryLoadFile(userFile) ?? tryLoadFile(packagePresetsFile()) ?? BUILTIN_PRESETS;
}

File: internal/presets/presets.go
// Package presets 是供应商目录（presets）加载层：用户覆盖文件 > 二进制旁路文件 > 内置兜底。
//
// 术语：一个保存的条目叫「配置」(profile，见 internal/config)，这里的目录条目叫「供应商」(provider)。
// 与 internal/config 解耦：presets 可以依赖 config（取存储目录），但 config 不依赖 presets。
//
// 实现说明：计划 §4.2 提到 go:embed，但 Go embed 无法引用包外的 ../../presets.json（npm 发布用的根文件）。
// 为避免在 internal/presets 再放一份同名副本造成漂移，这里改用与 TS 一致的「字面量 BuiltinPresets +
// 对拍测试（presets_test.go 断言它等于根 presets.json）」方案，根目录仍是唯一可编辑源。
package presets

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"

	"github.com/becomeless/cc-x/internal/config"
)

// 认证字段取值。
const (
	AuthToken  = "AUTH_TOKEN"
	AuthAPIKey = "API_KEY"
)

// URL 是某供应商的一个 API 地址（可多个，多个时让用户选）。
type URL struct {
	Label string `json:"label"`
	URL   string `json:"url"`
}

// Models 是四档模型映射（可部分为空）。
type Models struct {
	Opus   string `json:"opus,omitempty"`
	Sonnet string `json:"sonnet,omitempty"`
	Haiku  string `json:"haiku,omitempty"`
	Fable  string `json:"fable,omitempty"`
}

// Preset 是一个「供应商」目录条目。
type Preset struct {
	Name         string            `json:"name"`
	Auth         string            `json:"auth"` // AUTH_TOKEN | API_KEY
	URLs         []URL             `json:"urls"`
	Models       Models            `json:"models"`
	Effort       string            `json:"effort,omitempty"`
	Env          map[string]string `json:"env,omitempty"`
	ModelsAPI    string            `json:"models_api,omitempty"`     // 模型列表端点；缺省推导 {base}/v1/models（MiMo 路径不同需显式）
	ModelsAPIMap map[string]string `json:"models_api_map,omitempty"` // 按 BASE_URL 前缀匹配的模型列表端点（多地址供应商：按量/TokenPlan 各自端点）
	Models1M     []string          `json:"models_1m,omitempty"`      // 支持 1M 上下文的模型 ID 前缀表（[1m] 后缀默认开关）
}

// BuiltinPresets 是内置兜底目录，镜像仓库根 presets.json（由 presets_test.go 对拍保证不漂）。
// 第三方供应商绝不预置 `[1m]`（见 plan §3.1.1）。
var BuiltinPresets = []Preset{
	{
		Name:      "DeepSeek",
		Auth:      AuthToken,
		Effort:    "max",
		URLs:      []URL{{Label: "Anthropic 兼容", URL: "https://api.deepseek.com/anthropic"}},
		Models:    Models{Opus: "deepseek-v4-pro", Sonnet: "deepseek-v4-pro", Haiku: "deepseek-v4-flash"},
		ModelsAPI: "https://api.deepseek.com/models", // Anthropic 兼容端点未实现 /v1/models，模型列表在 OpenAI 风格端点
		Models1M:  []string{"deepseek-v4-pro"},
	},
	{
		Name:   "智谱GLM",
		Auth:   AuthToken,
		URLs:   []URL{{Label: "Anthropic 兼容", URL: "https://open.bigmodel.cn/api/anthropic"}},
		Models: Models{Opus: "GLM-4.7", Sonnet: "GLM-4.7", Haiku: "glm-4.5-air"},
		Env: map[string]string{
			"CLAUDE_CODE_DISABLE_NONESSENTIAL_TRAFFIC": "1",
		},
		ModelsAPI: "https://open.bigmodel.cn/api/anthropic/v1/models",
		Models1M:  []string{"glm-5.2"},
	},
	{
		Name: "小米MiMo",
		Auth: AuthToken,
		URLs: []URL{
			{Label: "按量付费API", URL: "https://api.xiaomimimo.com/anthropic"},
			{Label: "TokenPlan", URL: "https://token-plan-cn.xiaomimimo.com/anthropic"},
		},
		Models: Models{Opus: "mimo-v2.5-pro", Sonnet: "mimo-v2.5-pro", Haiku: "mimo-v2.5-pro"},
		// MiMo 的 models 端点是 OpenAI 风格 /v1/models（不带 /anthropic 前缀），按 base 前缀匹配各自端点
		ModelsAPIMap: map[string]string{
			"https://api.xiaomimimo.com/anthropic":           "https://api.xiaomimimo.com/v1/models",
			"https://token-plan-cn.xiaomimimo.com/anthropic": "https://token-plan-cn.xiaomimimo.com/v1/models",
		},
		Models1M: []string{"mimo-v2.5-pro"},
	},
	{
		Name:     "Kimi (月之暗面)",
		Auth:     AuthToken,
		Effort:   "max",
		URLs:     []URL{{Label: "Anthropic 兼容", URL: "https://api.moonshot.cn/anthropic"}},
		Models:   Models{Opus: "kimi-k3", Sonnet: "kimi-k3", Haiku: "kimi-k3", Fable: "kimi-k3"},
		Models1M: []string{"kimi-k3"},
	},
	{
		Name: "MiniMax",
		Auth: AuthToken,
		URLs: []URL{
			{Label: "国内站", URL: "https://api.minimaxi.com/anthropic"},
			{Label: "国际站", URL: "https://api.minimax.io/anthropic"},
		},
		Models:   Models{Opus: "MiniMax-M3", Sonnet: "MiniMax-M3", Haiku: "MiniMax-M3", Fable: "MiniMax-M3"},
		Models1M: []string{"MiniMax-M3"},
	},
	{
		Name:   "OpenRouter",
		Auth:   AuthAPIKey,
		URLs:   []URL{{Label: "Anthropic 兼容", URL: "https://openrouter.ai/api"}},
		Models: Models{Opus: "anthropic/claude-opus-5", Sonnet: "anthropic/claude-sonnet-5", Haiku: "anthropic/claude-haiku-4-5", Fable: "anthropic/claude-fable-5"},
	},
	{
		Name:   "百度千帆",
		Auth:   AuthToken,
		URLs:   []URL{{Label: "Coding Plan", URL: "https://qianfan.baidubce.com/anthropic/coding"}},
		Models: Models{Opus: "qianfan-code-latest", Sonnet: "qianfan-code-latest", Haiku: "qianfan-code-latest", Fable: "qianfan-code-latest"},
		Env: map[string]string{
			"CLAUDE_CODE_DISABLE_NONESSENTIAL_TRAFFIC": "1",
		},
	},
	{
		Name: "阿里百炼",
		Auth: AuthToken,
		URLs: []URL{
			{Label: "Token Plan（个人/团队）", URL: "https://token-plan.cn-beijing.maas.aliyuncs.com/apps/anthropic"},
			{Label: "按量付费", URL: "https://dashscope.aliyuncs.com/apps/anthropic"},
		},
		Models:   Models{Opus: "qwen3.8-max", Sonnet: "qwen3.8-max", Haiku: "qwen3.6-flash", Fable: "qwen3.8-max"},
		Models1M: []string{"qwen3.8-max", "qwen3.7-max", "qwen3.7-plus"},
	},
	{
		Name:   "火山方舟",
		Auth:   AuthToken,
		URLs:   []URL{{Label: "Coding Plan", URL: "https://ark.cn-beijing.volces.com/api/coding"}},
		Models: Models{Opus: "doubao-seed-2.0-code", Sonnet: "doubao-seed-2.0-code", Haiku: "doubao-seed-2.0-code", Fable: "doubao-seed-2.0-code"},
	},
	{
		Name:   "官方Anthropic",
		Auth:   AuthAPIKey,
		URLs:   []URL{{Label: "(留空，用登录态)", URL: ""}},
		Models: Models{},
	},
}

func asString(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}

func normalizeURL(raw any) URL {
	m, _ := raw.(map[string]any)
	return URL{Label: asString(m["label"]), URL: asString(m["url"])}
}

func normalizeModels(raw any) Models {
	m, _ := raw.(map[string]any)
	return Models{Opus: asString(m["opus"]), Sonnet: asString(m["sonnet"]), Haiku: asString(m["haiku"]), Fable: asString(m["fable"])}
}

func normalizeEnv(raw any) map[string]string {
	m, _ := raw.(map[string]any)
	allowed := []string{"CLAUDE_CODE_DISABLE_NONESSENTIAL_TRAFFIC", "CLAUDE_CODE_SUBAGENT_MODEL"}
	out := map[string]string{}
	for _, key := range allowed {
		if v, ok := m[key].(string); ok {
			out[key] = v
		}
	}
	if len(out) == 0 {
		return nil
	}
	return out
}

// normalizePreset 宽松规整一条；无名条目返回 ok=false 由调用方丢弃。
func normalizePreset(raw any) (Preset, bool) {
	m, _ := raw.(map[string]any)
	name := strings.TrimSpace(asString(m["name"]))
	if name == "" {
		return Preset{}, false
	}
	auth := AuthToken
	if asString(m["auth"]) == AuthAPIKey {
		auth = AuthAPIKey
	}
	urls := []URL{}
	if arr, ok := m["urls"].([]any); ok {
		for _, u := range arr {
			urls = append(urls, normalizeURL(u))
		}
	}
	p := Preset{Name: name, Auth: auth, URLs: urls, Models: normalizeModels(m["models"]), Env: normalizeEnv(m["env"])}
	if e := strings.TrimSpace(asString(m["effort"])); e != "" {
		p.Effort = e
	}
	p.ModelsAPI = strings.TrimSpace(asString(m["models_api"]))
	if mm, ok := m["models_api_map"].(map[string]any); ok && len(mm) > 0 {
		p.ModelsAPIMap = map[string]string{}
		for k, v := range mm {
			if s := strings.TrimSpace(asString(v)); s != "" {
				p.ModelsAPIMap[strings.TrimSpace(k)] = s
			}
		}
		if len(p.ModelsAPIMap) == 0 {
			p.ModelsAPIMap = nil
		}
	}
	if arr, ok := m["models_1m"].([]any); ok {
		for _, item := range arr {
			if s := strings.TrimSpace(asString(item)); s != "" {
				p.Models1M = append(p.Models1M, s)
			}
		}
	}
	return p, true
}

// normalizePresets 把任意解析结果规整为 []Preset；非数组或全空返回 nil（让调用方跌落兜底）。
func normalizePresets(raw any) []Preset {
	arr, ok := raw.([]any)
	if !ok {
		return nil
	}
	out := make([]Preset, 0, len(arr))
	for _, item := range arr {
		if p, ok := normalizePreset(item); ok {
			out = append(out, p)
		}
	}
	if len(out) == 0 {
		return nil
	}
	return out
}

// tryLoadFile 尝试读并解析一个 presets.json；任何问题都安静返回 nil。
func tryLoadFile(file string) []Preset {
	data, err := os.ReadFile(file)
	if err != nil {
		return nil
	}
	var raw any
	if json.Unmarshal(data, &raw) != nil {
		return nil
	}
	return normalizePresets(raw)
}

// Load 加载供应商目录。优先级：用户 <storeDir>/presets.json > 二进制旁路 presets.json > 内置 BuiltinPresets。
// 任一步缺失/损坏都安静跌落下一步，绝不中断启动。
func Load(storeDir string) []Preset {
	userFile := filepath.Join(config.ResolveStorePaths(storeDir).Dir, "presets.json")
	if p := tryLoadFile(userFile); p != nil {
		return p
	}
	if exe, err := os.Executable(); err == nil {
		if p := tryLoadFile(filepath.Join(filepath.Dir(exe), "presets.json")); p != nil {
			return p
		}
	}
	return BuiltinPresets
}
